### PR TITLE
*: remove project version 1 references, update migration guide to 2 -> 3

### PR DESCRIPTION
**Pull request by estroz**
Project version 1 shouldn't be mentioned in code, and any project version that isn't supported by the binary used in a project should cause an error to be returned.

This PR also adds a skeleton migration guide with some TODO's for future work that will likely be in before the next major release.

/kind documentation

### CODE REVIEW

**Comment by estroz** [comment]
Closes #1508

File: docs/book/src/SUMMARY.md [hunk 1]
```diff
@@ -45,12 +45,10 @@
 
 - [Migrations](./migrations.md)
 
-  - [Kubebuilder v1 vs v2](./migration/v1vsv2.md)
+  - [Kubebuilder v2 vs v3](./migration/v2vsv3.md)
```
Comment (camilamacedo86) [inline docs/book/src/SUMMARY.md:48]: Hi @estroz, 

IMO we should not remove the V1 to the V2 migration guide.  It should still
What do you need is add a new migration V2 to V3 instead.

File: build/test.sh [hunk 2]
```diff
@@ -1,43 +0,0 @@
...
-kubebuilder create resource --group insect --version v1beta1 --kind Wasp
-kubebuilder create controller --group apps --version v1beta2 --kind Deployment --core-type
-
-# Verify the controller-manager builds and the tests pass
-go build ./cmd/...
-go build ./pkg/...
-go test -race ./cmd/...
-go test -race ./pkg/...
-
-fi
```
Comment (camilamacedo86) [inline build/test.sh:43]: 👍 great

File: docs/book/src/migration/guide.md [hunk 3]
```diff
@@ -1,220 +1,58 @@
-# Migration from v1 to v2
+# Migration from v2 to v3
```
Comment (camilamacedo86) [inline docs/book/src/migration/guide.md:1]: Ditto above

File: docs/book/src/migration/multi-group.md [hunk 4]
```diff
@@ -1,85 +0,0 @@
-# Single Group to Multi-Group
```
Comment (camilamacedo86) [inline docs/book/src/migration/multi-group.md:1]: It should not be removed. 
For V2 and V# it still valid.

File: docs/book/src/migration/testdata/gopath/project-v1/.gitignore [hunk 5]
```diff
@@ -1,24 +0,0 @@
-
```
Comment (camilamacedo86) [inline docs/book/src/migration/testdata/gopath/project-v1/.gitignore:1]: I have a PR to remove the V1 mock and update the migration guide to not point to it. 
Could we move forward with it https://github.com/kubernetes-sigs/kubebuilder/pull/1499 before?

File: docs/book/src/migration/v2vsv3.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+The details of all changes (breaking or otherwise) can be found in
+[kubebuilder](https://github.com/kubernetes-sigs/kubebuilder/releases) release notes.
+
+## controller-tools
+
+#### CustomResourceDefinition-related
+
+The default `CustomResourceDefinition` version is now `v1`, bumped from `v1beta1` which is
+[deprecated][crds-deprecated-doc].
```
Comment (camilamacedo86) [inline docs/book/src/migration/v2vsv3.md:13]: was it merged?

File: docs/book/src/migration/v2vsv3.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+#### Webhook-related
+
+The default `ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` versions are now `v1`,
+bumped from `v1beta1` which is [deprecated][webhook-deprecated-pr].
```
Comment (camilamacedo86) [inline docs/book/src/migration/v2vsv3.md:18]: same ?

File: docs/kubebuilder_v0_v1_difference.md [hunk 8]
```diff
@@ -1,56 +0,0 @@
-# Kubebuilder v0 v.s. v1
```
Comment (camilamacedo86) [inline docs/kubebuilder_v0_v1_difference.md:1]: we should not remove the old docs. they should still.

File: docs/migration_guide.md [hunk 9]
```diff
@@ -1,146 +0,0 @@
-# Migration guide from v0 project to v1 project
```
Comment (camilamacedo86) [inline docs/migration_guide.md:1]: we should not remove the old docs. they should still.

File: internal/config/config.go [hunk 10]
```diff
@@ -64,7 +64,7 @@ func readFrom(fs afero.Fs, path string) (c config.Config, err error) {
 
 	// kubebuilder v1 omitted version, so default to v1
 	if c.Version == "" {
-		c.Version = config.Version1
+		return config.Config{}, fmt.Errorf("project version key `version` is empty or does not exist in %s", path)
```
Comment (camilamacedo86) [inline internal/config/config.go:67]: IMO we need a PR to ADD the migration guide v2 to v3
and another one to do these changes.

**Comment by estroz** [comment]
> we should not remove the old docs

You're probably right.

> the changes in the code needs be done in one Pr and we need to have another to add the new migration guide.

Sure

File: pkg/cli/init.go [hunk 11]
```diff
@@ -45,8 +45,8 @@ func (c *cli) newInitCmd() *cobra.Command {
 	// so that it shows up in help and does not cause a parse error.
 	cmd.Flags().String(projectVersionFlag, c.defaultProjectVersion,
 		fmt.Sprintf("project version, possible values: (%s)", strings.Join(c.getAvailableProjectVersions(), ", ")))
-	// The --plugins flag can only be called to init projects v2+.
-	if c.projectVersion != config.Version1 {
+	// The --plugins flag can only be called to init projects v3+.
+	if c.projectVersion != config.Version2 {
```
Comment (gabbifish) [inline pkg/cli/init.go:49]: Should this conditional continue to check `c.projectVersion != config.Version1`? Or in the case `v1` is used, explicitly print out a message saying that version is now deprecated? We could use the `IsVersionSupported()` helper to accomplish this.

File: pkg/model/config/config_test.go [hunk 12]
```diff
@@ -37,11 +37,6 @@ var _ = Describe("Config", func() {
 			expectedConfig Config
 		)
 
-		By("Using config version 1")
```
Comment (gabbifish) [inline pkg/model/config/config_test.go:40]: Should we keep this check to ensure that v1 usage continues to produce the expected behavior (failure?)

File: pkg/cli/init.go (hunk 11 above)
Comment (camilamacedo86) [reply pkg/cli/init.go:49]: V1 is no longer supported and its code no longer exists in the project.

**Comment by camilamacedo86** [comment]
Hi @estroz, 

In order to help we move forward and solve its blocker for the next release, I raise a PR with the migration guide only in  https://github.com/kubernetes-sigs/kubebuilder/pull/1714. Please, feel free to check. Also, feel free to change this pr to address only the other changes that you would like which are here.

File: pkg/cli/init.go (hunk 11 above)
Comment (camilamacedo86) [reply pkg/cli/init.go:49]: i agree that it needs to be IsVersionSupported()

Comment (camilamacedo86) [reply pkg/cli/init.go:49]: If we do this change here it then, shows that it will not work with v3+

**Comment by camilamacedo86** [comment]
Hi @estroz, 

The migration guide from V2 to V3 is merged. See: https://github.com/kubernetes-sigs/kubebuilder/pull/1714
Since it is open for I while I pushed your suggestions/fixes made here to remove the v1 and make the --plugins flag available only for V3-alpha project version via https://github.com/kubernetes-sigs/kubebuilder/pull/1763. It would be very nice to get your help in the review for we are able to merge it. 

in this way, I think we can close this. WDYT?

**Comment by camilamacedo86** [comment]
Hi @estroz, 

I am closing this one based on the comment  https://github.com/kubernetes-sigs/kubebuilder/pull/1636#issuecomment-720139141. However, please feel free to re-open if you see that it is required.

**Pull request closed without merging**